actors/market/src/lib.rs

```diff
@@ -543,137 +550,226 @@ impl Actor {
+                        verified_infos.push(VerifiedDealInfo {
+                            client: proposal.client.id().unwrap(),
+                            allocation_id: alloc_id,
+                            data: proposal.piece_cid,
+                            size: proposal.piece_size,
+                        })
+                    } else {
+                        nonverified_deal_space += proposal.piece_size.0;
                     }
-                };
 
-                sector_deals.push((p.sector_number, SectorDealIDs { deals: p.deal_ids.clone() }));
+                    deal_states.push((
+                        *deal_id,
+                        DealState {
+                            sector_number: p.sector_number,
+                            sector_start_epoch: curr_epoch,
+                            last_updated_epoch: EPOCH_UNDEFINED,
+                            slash_epoch: EPOCH_UNDEFINED,
+                            verified_claim: alloc_id,
+                        },
+                    ));
+                }
+                for id in &p.deal_ids {
+                    activated_deals.insert(*id);
+                }
 
-                // Update deal states
-                let mut verified_infos = Vec::new();
+                sectors_deals.push((p.sector_number, SectorDealIDs { deals: p.deal_ids.clone() }));
+                activations.push(SectorDealActivation { nonverified_deal_space, verified_infos });
+                batch_gen.add_success();
+            }
 
-                // This construction could be replaced with a single "update deal state"
-                // state method, possibly batched over all deal ids at once.
-                let update_result: Result<(), ActorError> =
-                    proposals.into_iter().try_for_each(|(deal_id, proposal)| {
-                        let s = st
-                            .find_deal_state(rt.store(), deal_id)
-                            .context(format!("error looking up deal state for {}", deal_id))?;
+            st.put_deal_states(rt.store(), &deal_states)?;
+            st.put_sector_deal_ids(rt.store(), &miner_addr, &sectors_deals)?;
+            st.save_pending_deal_allocation_ids(&mut pending_deal_allocation_ids)?;
+            Ok((activations, batch_gen.gen()))
+        })?;
 
-                        if s.is_some() {
-                            return Err(actor_error!(
-                                illegal_argument,
-                                "deal {} already activated",
-                                deal_id
-                            ));
-                        }
+        Ok(BatchActivateDealsResult { activations, activation_results: batch_ret })
+    }
 
-                        let propc = rt_deal_cid(rt, &proposal)?;
+    /// Receives notification of a change to sector content, which may satisfy to activate a deal.
+    /// Deals are activated or fail independently, including in the same sector.
```

Comment:
I think the property that makes this ok is that deal ids are not tied to sector info in this case.  If we follow through on removing them entirely then we could possibly modify builtin batch to have the same semantics.

Reply:
We could, but this would be a change in behaviour for SPs which I think we probably shouldn't enact if we can avoid it.